File: binary2newick matlab/binary2newick.m
function output = binary2newick(inputPath, varargin)

% This script reads a binary tree matrix from an excel file and converts
% it into the newick format. The excel file must not contain anything but the matrix.
% The first argument specifys the input path. If
% no other arguments are used, the script writes the result to the standard
% output (newick output.txt) and overwrites, if the file already exists. To
% change the output directory, use a second argument to specify the output
% directory and a third argument to set the writing mode ('w': overwrite,
% 'at': append to existing file).

outputPath = 'newick output.txt';
mode = 'w';
if ~isempty(varargin)
    outputPath = varargin{1};
    mode = varargin{2};
end

    [num,txt,raw] = xlsread(inputPath);
    for i = 1:size(num, 1)
        subTrees{i} = 'e';
    end
    
    IndexStart = 1;
    
    for i = 1:size(num, 2)
        buffer = {};
        bufferIndex = 1;
        check = true;
        for j = 1:size(num, 1)
            if num(j, i) == 1
                if check
                    IndexStart = j;
                    check = false;
                end
                if ~strcmp(subTrees{j}, 'e') && ~strcmp('x', subTrees{j})
                    buffer(bufferIndex) = subTrees(j);
                    bufferIndex = bufferIndex+1;
                    subTrees(j) = cellstr('x');
                elseif strcmp(subTrees{j}, 'e')
                    buffer(bufferIndex) = cellstr(strrep(txt{j}, ' ', '_'));
                    bufferIndex = bufferIndex+1;
                    subTrees(j) = cellstr('x');
                end
            end
        end
        subTrees(IndexStart) = cellstr(createClade(buffer));
    end
    subTrees(1) = cellstr(strcat(subTrees{1}, ';'));
    output = subTrees(1);
    fid  = fopen(outputPath,mode);
    fprintf(fid, '%s\n', subTrees{1});
    fclose(fid);
    
end


%This function takes a cell arresultray of subtree strings and returns them as
%concatinated string in format of a newick clade.
function output = createClade(input)
    output = strcat('(', input{1});
    for i=2:1:size(input, 2)
        output = strcat(output, ',', input{i});
    end
    output = strcat(output, ')');
end

File: binary2newick.py
import sys
from openpyxl import load_workbook

#This script reads the binaray matrix for a phylogenetic tree from an excel file and converts it into newick format.
#The first argument specifys the input directory. The second and third argument is optional and define the output directory
#and the writing mode (use "w" in the third argument to overwrite and "a" to append to existing file).

#Arguments: inputDir <outputDir> <mode>

wb = load_workbook(sys.argv[1])
ws = wb.active

outputDir = 'newick output.txt'
mode = 'w'
IndexStart = 0
cLength = 0
rLength = 0

if len(sys.argv) > 2:
    outputDir = sys.argv[2]
if len(sys.argv) > 3:
    mode = sys.argv[3]

for row in ws.columns[0]:
    cLength += 1

for col in ws.rows[0]:
    rLength += 1

subTrees = ['e' for stuff in range(cLength)]


def create_clade(inp):
    result = '(' + inp[0]
    for element in inp[1:]:
        if element == '0':
            break
        result = result + ',' + element
    return result + ')'


i = 0
while i < cLength-2:
    buffer = ['0' for data in range(cLength)]
    bufferIndex = 0
    check = True
    j = 0
    while j < rLength+1:
        if ws.cell(row=j+1, column=i+2).value == 1:
            if check:
                IndexStart = j
                check = False
            if subTrees[j] != 'e' and subTrees[j] != 'x':
                buffer[bufferIndex] = subTrees[j]
                bufferIndex += 1
                subTrees[j] = 'x'
            elif subTrees[j] == 'e':
                valueString = ws.cell(row=j+1, column=1).value
                buffer[bufferIndex] = valueString.replace(' ', '_')
                bufferIndex += 1
                subTrees[j] = 'x'
        j += 1
    subTrees[IndexStart] = create_clade(buffer)
    i += 1
subTrees[0] += ';'
print(subTrees[0])

fobj_out = open(outputDir, mode)
fobj_out.write(subTrees[0] + '\n')
fobj_out.close()
